app/services/keys/claim_service.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+class Keys::ClaimService < BaseService
+  HEADERS = { 'Content-Type' => 'application/activity+json' }.freeze
+
+  class Result < ActiveModelSerializers::Model
+    attributes :account, :device_id, :key_id,
+               :key, :signature
+
+    def initialize(account, device_id, key_attributes = {})
+      @account   = account
+      @device_id = device_id
+      @key_id    = key_attributes[:key_id]
+      @key       = key_attributes[:key]
+      @signature = key_attributes[:signature]
+    end
+  end
+
+  def call(account_id, device_id)
+    @account   = Account.find(account_id)
+    @device_id = device_id
+
+    if @account.local?
+      claim_local_key!
+    else
+      claim_remote_key!
+    end
+  rescue ActiveRecord::RecordNotFound
+    nil
+  end
+
+  private
+
+  def claim_local_key!
+    device = @account.devices.find_by(device_id: @device_id)
+    key    = nil
+
+    ApplicationRecord.transaction do
+      key = device.one_time_keys.order(Arel.sql('random()')).first!
+      key.destroy!
+    end
+
+    @result = Result.new(@account, @device_id, key)
+  end
+
+  def claim_remote_key!
+    query_result = QueryService.new.call(@account)
+    device       = query_result.find(@device_id)
+
+    return unless device.present? && device.valid_claim_url?
+
+    json = fetch_resource_with_post(device.claim_url)
+
+    return unless json.present? && json['publicKeyBase64'].present?
+
+    @result = Result.new(@account, @device_id, key_id: json['id'], key: json['publicKeyBase64'], signature: json.dig('signature', 'signatureValue'))
+  end
+
+  def fetch_resource_with_post(uri)
+    build_post_request(uri).perform do |response|
+      raise Mastodon::UnexpectedResponseError, response unless response_successful?(response) || response_error_unsalvageable?(response)
+
+      body_to_json(response.body_with_limit) if response.code == 200
+    end
+  end
+
+  def build_post_request(uri)
+    Request.new(:post, uri).tap do |request|
+      request.on_behalf_of(Account.representative, :uri)
```

Comment:
I wonder if there's any benefit in hiding who is claiming the key here. Benefits of not doing that would be the receiving server being able to preemptively reject giving out keys to people the receiver doesn't want messages from (e.g., blocked peeps, or anyone not a mutual if we introduce such a limitation, etc.) and thus reduce potential for key pool exhaustion. It would also allow the receiving server to do rate-limiting based on the account rather than the instance. I'm not sure what the privacy implications would be, since normal use would have the client submit a message right away anyway.

Reply:
Signing on behalf of the account doing the claiming is something I wanted to do anyways.